File: native/remote-host-app/Sources/RemoteHostRuntime/Bootstrap/NativeProtectedModelOperation.swift
import Foundation

/** One sealed provider/model route selected by the signed Host build. */
public struct NativeProtectedModelRoute: Equatable, Sendable {
  public let provider: String
  public let model: String

  private init(provider: String, model: String) {
    self.provider = provider
    self.model = model
  }

  /** The only shipped route until an explicitly reviewed build changes it. */
  public static let sealedDefault = NativeProtectedModelRoute(provider: "deepseek-official", model: "deepseek-v4-flash")
}

/** Text-only admitted work for the Keychain-retaining native model operation. */
public struct NativeProtectedModelTurn: Equatable, Sendable {
  public let sessionID: String
  public let text: String

  public init(sessionID: String, text: String) {
    self.sessionID = sessionID
    self.text = text
  }
}

/**
 A signed-Host-only model operation. Its implementation retains any Keychain
 credential and exposes neither credential bytes nor a generic provider,
 endpoint, or token lookup API to the Node child or FD198.
 */
public protocol NativeProtectedModelOperation: Sendable {
  func complete(_ turn: NativeProtectedModelTurn) async throws -> String
}

/**
 The native registration point for an already-provisioned protected model
 operation. This source deliberately does not install, inspect, or request a
 credential; provisioning remains an explicit signed-Host action.
 */
public protocol NativeProtectedModelOperationProvider: Sendable {
  func protectedModelOperation() throws -> any NativeProtectedModelOperation
}

File: native/remote-host-app/GatewayRuntime/native-protected-model-operation.ts
/**
 * Typed native operation for one fixed signed-Host model route. It deliberately
 * accepts work and returns model output, never credential material or a
 * generic provider client. The production native implementation is not linked
 * yet; this is the only permitted seam for its future Keychain use.
 */

/** Fixed provider/model identity chosen by the signed Host build. */
export interface FixedModelIdentity {
  readonly provider: string
  readonly model: string
}

/** One text-only user turn supplied to the protected model operation. */
export interface ProtectedModelTurn {
  readonly sessionId: string
  readonly model: FixedModelIdentity
  readonly text: string
}

/** A single approval requested by the model's fixed agent execution. */
export interface ProtectedModelApprovalRequest {
  readonly sessionId: string
  readonly toolName: string
  readonly reason?: string
}

/** A single question requested by the model's fixed agent execution. */
export interface ProtectedModelQuestionRequest {
  readonly sessionId: string
  readonly questions: readonly { readonly id: string; readonly question: string }[]
}

/** The model operation's interaction callbacks, implemented by the sealed core. */
export interface ProtectedModelInteractions {
  requestApproval(request: ProtectedModelApprovalRequest): Promise<'allowed-once' | 'rejected'>
  requestQuestion(request: ProtectedModelQuestionRequest): Promise<unknown>
}

/** A completed model turn. Streaming belongs inside the native protected operation. */
export interface ProtectedModelCompletion { readonly text: string }

/**
 * Native Keychain-only model operation. It has no credential getter, export,
 * arbitrary endpoint parameter, or phone-controlled input beyond one admitted
 * text turn. A future signed Host implementation must retain Keychain access
 * in native code and expose only this operation to the Node child.
 */
export interface NativeProtectedModelOperation {
  readonly capability: 'signed-host-keychain-model-operation-v1'
  readonly model: FixedModelIdentity
  complete(turn: ProtectedModelTurn, interactions: ProtectedModelInteractions, signal: AbortSignal): Promise<ProtectedModelCompletion>
}
